### ISSUE TITLE
Status of PSK Support?

### ISSUE THREAD

**Issue by stevedonovan** [ctz/rustls#174]
Just curious, since it has been on the wishlist for a long time.  Very useful feature when you deploy your own boxes in someone else's infrastructure.

**Comment by ctz** [comment]
There's been no work on this as yet.

**Comment by TotalKrill** [comment]
Do you have any outline to what would be required, or how to get this ball rolling? I understand priorities and and that time is always a limitation. 

I came here since a project that is using this nice library would benefit from having this as a connection option.

**Comment by djc** [comment]
I recently talked to a potential client who wanted to fund this work, but they decided to postpone it for now. Nonetheless, if anyone wants to fund the work, I'd be happy to talk. (Otherwise, I'm afraid I won't be able to make time for this.)

**Comment by TotalKrill** [comment]
Hello again, 

I am revisiting this issue again. Guessing no progress has been done yet. @djc, is this something that could be crowdfunded or something?

**Comment by djc** [comment]
I'm not sure if there are any good platforms to manage this kind of thing. However, if you want to send me an email pledging some amount of funding, I can make an estimate of the amount of work and see when enough funding has been pledged.

**Comment by TotalKrill** [comment]
@djc I sent a mail

**Comment by TotalKrill** [comment]
I sent a response to your email, maybe check the spam?

feeling a bit uncomfortable using this issue tracker as a chat though @djc

**Comment by dignifiedquire** [comment]
I am looking for support of https://datatracker.ietf.org/doc/html/rfc8773 is this what this issue is about, or should I create a new issue for that?

**Comment by est31** [comment]
That RFC is different from the PSK support that is part of RFC8446 directly. See the section in RFC8773:

> The TLS 1.3 [[RFC8446](https://datatracker.ietf.org/doc/html/rfc8446)] handshake protocol provides two mutually exclusive forms of server authentication. First, the server can be authenticated by providing a signature certificate and creating a valid digital signature to demonstrate that it possesses the corresponding private key. Second, the server can be authenticated by demonstrating that it possesses a pre-shared key (PSK) that was established by a previous handshake. A PSK that is established in this fashion is called a resumption PSK. A PSK that is established by any other means is called an external PSK. This document specifies a TLS 1.3 extension permitting certificate-based server authentication to be combined with an external PSK as an input to the TLS 1.3 key schedule.

**Comment by dtoubelis** [comment]
I'm very much interested in this feature. We are working on post-quantum encryption solutions and they are all based on symmetric keys and having TLS-PSK available in Rust would be a great help.

**Comment by fvp** [comment]
Hi, I am interested in helping to implement PSK support in RUSTLS. Should I just send pull requests or is there some kind of architecture, rules that I should follow?

**Comment by djc** [comment]
@fvp as this is not part of the project's direct priorities, it would be a good idea to try to do some design upfront. I would suggest:

* Read [`CONTRIBUTING.md`](https://github.com/rustls/rustls/blob/main/CONTRIBUTING.md)
* Research the code base and try to figure out a design that makes sense
* Post the design here to gather feedback

Only once consensus on the design has been reached, it would make sense to move forward with implementation.

**Comment by happytraveller-alone** [comment]
Does the current project have any implementation of PSK (Pre-Shared Key)? If so, could you expose an interface for PSK verification? This would be extremely useful for research into the handshake process. Additionally, any documentation or guidance on how to integrate and utilize PSK within the handshake would be greatly appreciated.

**Comment by cpu** [comment]
> Does the current project have any implementation of PSK (Pre-Shared Key)?

It does not. See [djc's reply](https://github.com/rustls/rustls/issues/174#issuecomment-1891775597) for next steps.

**Comment by djc** [comment]
As in 2021, I am available if you want to fund this work.

**Comment by elagergren-spideroak** [comment]
SpiderOak is currently working on adding TLS 1.3 external PSK support to rustls to support our own needs. We plan to upstream the code.

You can track progress here: https://github.com/elagergren-spideroak/rustls/pull/1

**Comment by djc** [comment]
I investigated this a bunch last year, here's what I wrote up:

> At least some TLS implementations (notably BoringSSL) don't implement PSK support for TLS 1.3. This also means that the bogo test suite that we use to flush out interoperability issues does not support testing PSK support for TLS 1.3. Google and Cloudflare have since published RFC 9258 which provides a different approach to PSKs in TLS 1.3, and for rustls we'd probably prefer to stick with that specification.
> 
> We are also worried about the risk of replay attacks (see https://datatracker.ietf.org/doc/html/rfc8446#section-8) for early
data support in combination with long-lived PSKs. We currently only support single-use tickets, and the alternatives (8.2, client hello recording, and 8.3, freshness checks) don't seem attractive to us.

**Comment by TaeHagen** [comment]
Apparently, to talk to quickrelay (for FaceTime), Apple requires TLS-PSK support. I got it working with the PR'd branch, I had to make a few tweaks to get it to work, main one is using "ext binder" instead of "res binder" as the HKDF tag.

I'd be willing to fund getting this ported to the latest version of rustls and upstreamed. Alternatively, I could work on porting it myself, however, given it is a substantial changeset and you have opinionated standards (which is fair, it is your project after all), it might just be easier to pay for a few days of your time @djc @ctz. Please let me know how you want to proceed.

**Comment by TaeHagen** [comment]
Any updates on this?

**Comment by djc** [comment]
Thanks for your offer to fund this work. Unfortunately we have fairly limited capacity right now and we think releasing 0.24.0 is higher priority. We're tentatively planning for 0.24.0 to be published around the end of July, so I suggest you contact us again around that time to discuss getting PSK support merged.

**Comment by TaeHagen** [comment]
I'm still willing to fund getting PSK support merged. It seems you are still working on 0.24, so when you have capacity to get this merged please let me know.
